sdk/tools/pefixup.c

```diff
@@ -121,26 +127,113 @@ static int add_loadconfig(unsigned char *buffer, PIMAGE_NT_HEADERS nt_header)
     return 1;
 }
 
+static int driver_fixup(int mode, unsigned char *buffer, PIMAGE_NT_HEADERS nt_header)
+{
+    /* GNU LD just doesn't know what a driver is, and has notably no idea of paged vs non-paged sections */
+    for (unsigned i = 0; i < nt_header->FileHeader.NumberOfSections; i++)
+    {
+        PIMAGE_SECTION_HEADER Section = IMAGE_FIRST_SECTION(nt_header) + i;
+
+        /* LD puts alignment crap that nobody asked for */
+        Section->Characteristics &= ~IMAGE_SCN_ALIGN_MASK;
+
+        /* LD overdoes it and puts the initialized flag everywhere */
+        if (Section->Characteristics & IMAGE_SCN_CNT_CODE)
+            Section->Characteristics &= ~IMAGE_SCN_CNT_INITIALIZED_DATA;
+
+        /* For some reason, .rsrc is made writable by windres */
+        if (strncasecmp((char*)Section->Name, ".rsrc", 5) == 0)
+        {
+            Section->Characteristics &= ~IMAGE_SCN_MEM_WRITE;
+            continue;
+        }
+
+        /* Known sections which can be discarded */
+        if (strncasecmp((char*)Section->Name, "INIT", 4) == 0)
+        {
+            Section->Characteristics |= IMAGE_SCN_MEM_DISCARDABLE;
+            continue;
+        }
+
+        /* Known sections which can be paged */
+        if ((strncasecmp((char*)Section->Name, "PAGE", 4) == 0)
+            || (strncasecmp((char*)Section->Name, ".rsrc", 5) == 0)
+            || (strncasecmp((char*)Section->Name, ".edata", 6) == 0)
+            || (strncasecmp((char*)Section->Name, ".reloc", 6) == 0))
+        {
+            continue;
+        }
+
+        /* If it's discardable, don't set the flag */
+        if (Section->Characteristics & IMAGE_SCN_MEM_DISCARDABLE)
+            continue;
+
+        Section->Characteristics |= IMAGE_SCN_MEM_NOT_PAGED;
+    }
+
+    return 0;
+}
+
+static
+void
+print_usage(void)
+{
+    printf("Usage: %s <mode> <filename>\n", g_ApplicationName);
+    printf("Where <mode> is on of the following:\n");
+    printf("  --loadconfig:         Fix the LOAD_CONFIG directory entry\n");
+    printf("  --kernelmodedriver:   Fix code and data sections for driver images\n");
+    printf("  --wdmdriver:          Fix code and data sections for WDM drivers\n");
+    printf("  --kerneldll:          Fix code and data sections for Kernel-Mode DLLs\n");
+    printf("  --kernel:             Fix code and data sections for kernels\n");
```

Comment:
```suggestion
    printf("  --loadconfig:         Fix the LOAD_CONFIG directory entry\n");
    printf("  --kernelmodedriver    Fix code and data sections for driver images\n");
    printf("  --wdmdriver           Fix code and data sections for WDM drivers\n");
    printf("  --kerneldll           Fix code and data sections for Kernel-Mode DLLs\n");
    printf("  --kernel              Fix code and data sections for kernels\n");
```
(Since the options do not require a semicolon following them for working.)